csrc/scheduler/mma_utils.cpp

```diff
@@ -1229,6 +1229,86 @@ RolesMapOpt getTensorsRoles(Fusion* fusion) {
   return roles_map;
 }
 
+namespace {
+
+void add_mma_op_helper(TensorView* tv_a, TensorView* tv_b, TensorView* out) {
+  auto* init = IrBuilder::create<Val>(0.0);
+
+  NVF_CHECK(
+      tv_a->getDataType().value() == DataType::Half ||
+      tv_a->getDataType().value() == DataType::BFloat16);
+  NVF_CHECK(tv_a->getDataType().value() == tv_b->getDataType().value());
+
+  IrBuilder::create<MmaOp>(out, tv_a, tv_b, init);
+}
+
+TensorView* get_tensorview_prior_to_cast(TensorView* in) {
+  if (in->getDataType() == DataType::BFloat16 ||
+      in->getDataType() == DataType::Half) {
+    return in;
+  }
+  TensorView* ret = nullptr;
+  if (in->getDataType() == DataType::Float) {
+    // get the producer.
+    auto* cast = in->definition();
+    auto* uCastOp = dynamic_cast<UnaryOp*>(cast);
+    if (uCastOp && uCastOp->getUnaryOpType() == UnaryOpType::Cast) {
+      auto* aTV = static_cast<TensorView*>(uCastOp->in());
+      NVF_CHECK(
+          (aTV->getDataType() == DataType::BFloat16 ||
+           aTV->getDataType() == DataType::Half),
+          "The output of the cast must be half or BF16");
+      ret = aTV;
+    }
+  }
+  return ret;
+}
+
+void add_mma_op(
+    Fusion* fusion,
+    std::vector<std::pair<ReductionOp*, BinaryOp*>>& sum_mul_pairs) {
+  for (auto [sumOp, mulOp] : sum_mul_pairs) {
+    // The inputs to mul may be cast to a float. These helpers get the
+    // tensorviews prior to the cast.
+    auto x =
+        get_tensorview_prior_to_cast(static_cast<TensorView*>(mulOp->lhs()));
+    auto y =
+        get_tensorview_prior_to_cast(static_cast<TensorView*>(mulOp->rhs()));
+    add_mma_op_helper(x, y, static_cast<TensorView*>(sumOp->output(0)));
+  }
+};
+} // namespace
+
+
+void CombineMulSum::handle(ReductionOp* stmt) {
+  // Check if operation is a sum.
+  if (stmt->getReductionOpType() == BinaryOpType::Add) {
+    auto* inputOfSum = stmt->in();
+    if (inputOfSum != nullptr) {
+      auto* producer = inputOfSum->definition();
+      // Then check if the prodcer of the sum is a mul.
+      if (producer && dynamic_cast<BinaryOp*>(producer)) {
+        auto* bOp = dynamic_cast<BinaryOp*>(producer);
+        // If it'a mul followed by a sum, put this in a list.
+        if (bOp->getBinaryOpType() == BinaryOpType::Mul) {
```

Comment:
Should we actually check the domains of these tensors and only add it to `sum_mul_pairs_` if it is really a matmul? Not all mul-sum are matmuls, it could be just a pointwise mul followed by a sum.

Reply:
@naoyam said something similar. I am working on that.

Reply:
Thanks, I added checks, please take a look.

Reply:
Shouldn't we also check the two inputs should have a broadcast in a different position, and the reduction domain should be the common non-broadcast domain?